birdhouse/read-configs.include.sh

```diff
@@ -58,38 +58,46 @@ discover_compose_dir() {
         fi
         export COMPOSE_DIR
     fi
+    # Perform last-chance validation in case 'COMPOSE_DIR' was incorrectly set explicitly
+    # and that 'read-configs.include.sh' was sourced directly from an invalid location.
+    if [ ! -f "${COMPOSE_DIR}/pavics-compose.sh" ]; then
+        echo \
+          "CRITICAL: [${COMPOSE_DIR}/pavics-compose.sh] not found," \
+          "please set variable 'COMPOSE_DIR' to a valid location." \
+          "Many features depend on this variable." 1>&2
+        return 2
+    fi
 }
 
 
-discover_compose_dir
-if [ -f "${COMPOSE_DIR}/scripts/logging.include.sh" ]; then
-    . "${COMPOSE_DIR}/scripts/logging.include.sh"
-fi
+# error out appropriately without closing shell according to 'sh <script>' or '. <script>' call
+discover_compose_dir || return $? 2>/dev/null || exit $?
```

Comment:
Why `2>/dev/null`?  This will mask the CRITICAL message that we actually want to report.

Reply:
No. Either way, the echo'd message in `discover_compose_dir` is printed before return 2.
The `2>/dev/null` is to avoid some shell's message that `return` cannot be done outside a function, in which case the failure is piped by `exit` instead.

Reply:
The file descriptor redirection only applies to the return statement, not any output from `discover_compose_dir`.

Reply:
I usually do `2>/dev/null` to silence STDERR.  But above, in `discover_compose_dir` you did 
```sh
        echo \
          "CRITICAL: [${COMPOSE_DIR}/pavics-compose.sh] not found," \
          "please set variable 'COMPOSE_DIR' to a valid location." \
          "Many features depend on this variable." 1>&2
```
which means redirect STDOUT to STDERR.

So the combination of both changes looks to me that the CRITICAL message will be silenced.  Let me know if that is indeed your intention or I might have understood the syntax wrong.

Reply:
The `2>/dev/null` does not affect any of the output from the `discover_compose_dir` function. It will only redirect stderr after the `||`.

For example @tlvu, run this in your shell:

```
echo 'error message' 1>&2 && return 1 2>/dev/null
```

and you'll see that 'error' message is still visible.

Reply:
Like @mishaschwartz said, each redirect is independent.
For them to chain, you would need something like `( cmd1 1>&2 || cmd2 1>&2 ) 2>/dev/null`.

Reply:
Oh right, it's true, it's not inside a parenthesis.  Sorry, not sure why my brain was adding an imaginary parenthesis !